Firestore/core/src/firebase/firestore/remote/stream.mm

```diff
@@ -0,0 +1,318 @@
+using util::StringFormat;
+
+namespace {
+
+/**
+ * Initial backoff time after an error.
+ * Set to 1s according to https://cloud.google.com/apis/design/errors.
+ */
+const double kBackoffFactor = 1.5;
+const AsyncQueue::Milliseconds kBackoffInitialDelay{std::chrono::seconds(1)};
+const AsyncQueue::Milliseconds kBackoffMaxDelay{std::chrono::seconds(60)};
+/** The time a stream stays open after it is marked idle. */
+const AsyncQueue::Milliseconds kIdleTimeout{std::chrono::seconds(60)};
+
+}  // namespace
+
+Stream::Stream(AsyncQueue* worker_queue,
+               CredentialsProvider* credentials_provider,
+               Datastore* datastore,
+               TimerId backoff_timer_id,
+               TimerId idle_timer_id)
+    : worker_queue_{worker_queue},
+      credentials_provider_{credentials_provider},
+      datastore_{datastore},
+      backoff_{worker_queue, backoff_timer_id, kBackoffFactor,
+               kBackoffInitialDelay, kBackoffMaxDelay},
+      idle_timer_id_{idle_timer_id} {
+}
+
+// Starting
+
+void Stream::Start() {
+  EnsureOnQueue();
+
+  if (state_ == State::Error) {
+    BackoffAndTryRestarting();
+    return;
+  }
+
+  LOG_DEBUG("%s start", GetDebugDescription());
+
+  HARD_ASSERT(state_ == State::Initial, "Already started");
+  state_ = State::Starting;
+
+  Authenticate();
+}
+
+void Stream::Authenticate() {
+  EnsureOnQueue();
+
+  // Auth may outlive the stream, so make sure it doesn't try to access a
+  // deleted object.
+  std::weak_ptr<Stream> weak_this{shared_from_this()};
+  int auth_generation = generation_;
+  credentials_provider_->GetToken(
+      [weak_this, auth_generation](StatusOr<Token> maybe_token) {
+        auto strong_this = weak_this.lock();
+        if (!strong_this) {
+          return;
+        }
+        strong_this->worker_queue_->EnqueueRelaxed(
+            [maybe_token, weak_this, auth_generation] {
+              auto strong_this = weak_this.lock();
+              // Streams can be stopped while waiting for authorization, so need
+              // to check generation.
+              if (!strong_this || strong_this->generation_ != auth_generation) {
+                return;
+              }
+              strong_this->ResumeStartAfterAuth(maybe_token);
+            });
+      });
+}
+
+void Stream::ResumeStartAfterAuth(const StatusOr<Token>& maybe_token) {
+  EnsureOnQueue();
+
+  HARD_ASSERT(state_ == State::Starting,
+              "State should still be 'Starting' (was %s)", state_);
+
+  if (!maybe_token.ok()) {
+    OnStreamError(maybe_token.status());
+    return;
+  }
+
+  absl::string_view token = [&] {
+    auto token = maybe_token.ValueOrDie();
+    return token.user().is_authenticated() ? token.token()
+                                           : absl::string_view{};
+  }();
+
+  grpc_stream_ = CreateGrpcStream(datastore_, token);
+  grpc_stream_->Start();
+}
+
+void Stream::OnStreamStart() {
+  EnsureOnQueue();
+
+  state_ = State::Open;
+  DoOnStreamStart();
```

Comment:
The `Do` prefix is supposed to imply that this is a customization point for derived classes.

I'm a little torn here. On one hand, I'd prefer for `Stream` to be entirely unaware of delegates, which won't really be the case if it used "notify" (I'd be happy with the tradeoff if derived streams could be unaware of gRPC streams instead, but that doesn't appear to be possible). On the other hand, it seems useful to separate the terminology, so that "start" and "finish" only apply to gRPC streams, while "open" and "close" only apply to Firestore streams.

Since I don't have a strong opinion on this, changed to `NotifyStreamOpen`/`NotifyStreamResponse`/`NotifyStreamClose`. Feel free to tweak the names.